plonk/src/circuit/basic.rs

```diff
@@ -1490,100 +1500,85 @@ impl<F: PrimeField> PlonkCircuit<F> {
         let cmp_result = self.logic_and(msb_eq, cmp_result)?;
 
         self.logic_or_gate(msb_check, cmp_result)
     }
 
-    /// Helper function to check whether `a` is no greater
-    /// than a given constant. Let N = F::size_in_bits()
-    /// It uses N negation gates and N AND/OR gates.
-    fn is_leq_constant_internal(
+    /// Helper function to check whether `a` is greater than a given
+    /// constant. Let N = F::size_in_bits(), it assumes that the
+    /// constant < 2^N. And it uses at most N AND/OR gates.
+    fn is_ge_constant_internal(
         &mut self,
         a: Variable,
         constant: &F,
     ) -> Result<BoolVar, PlonkError> {
         let a_bits_le = self.unpack(a, F::size_in_bits())?;
         let const_bits_le = constant.into_repr().to_bits_le();
-        let bits_length = const_bits_le.len().max(a_bits_le.len());
-
-        // Stack for constructing the comparison gate
-        let mut current_run = Vec::new();
-        // Iterate from LSB to MSB. If current bit of constant is 0
-        // push a NOT(a) to the stack. Otherwise, wrap and clear the
-        // current stack using an k-ary AND gate, OR the result with
-        // a NOT(a) and push it back to the stack.
+
+        // Iterating from LSB to MSB. Skip the front consecutive 1's.
+        // Put an OR gate for bit 0 and an AND gate for bit 1.
         const_bits_le
             .into_iter()
-            .take(bits_length)
-            .zip(a_bits_le.iter().take(bits_length))
-            .try_for_each(|(b, a)| -> Result<(), PlonkError> {
-                if b {
-                    let c = self.logic_neg(*a)?;
-                    let d = self.logic_and_all(&current_run)?;
-                    let c = self.logic_or(c, d)?;
-                    current_run.clear();
-                    current_run.push(c);
-                } else {
-                    let c = self.logic_neg(*a)?;
-                    current_run.push(c);
-                }
-                Ok(())
-            })?;
-        // If bits length of `a` is too long, check whether the most significant
-        // bits of `a` is 0.
-        a_bits_le
-            .iter()
-            .skip(bits_length)
-            .try_for_each(|a| -> Result<(), PlonkError> {
-                let c = self.logic_neg(*a)?;
-                current_run.push(c);
-                Ok(())
-            })?;
-        if current_run.len() == 1 {
-            Ok(current_run[0])
-        } else {
-            self.logic_and_all(&current_run)
-        }
+            .chain(ark_std::iter::repeat(false))
+            .take(a_bits_le.len())
+            .zip(a_bits_le.iter())
+            .skip_while(|(b, _)| *b)
+            .try_fold(
+                BoolVar(self.one()),
+                |current, (b, a)| -> Result<BoolVar, PlonkError> {
+                    if b {
+                        self.logic_and(*a, current)
+                    } else {
+                        if current.0 == self.one() {
```

Comment:
This circuit implementation is input/witness dependent.

Reply:
I don't understand why it's witness dependent here. `b` here is from some constant.

Reply:
It seems that when b == 0, the gate depends on the value of `current.0`, or did I miss sth?

Reply:
`current` is a BoolVar. So `current.0` is just a variable. It's just checking whether `current` is the initial value `BoolVar(self.one())`

Reply:
Oh you're right, this line will only be triggered at the beginning. But for readability, maybe we can directly write `self.logic_or()`.